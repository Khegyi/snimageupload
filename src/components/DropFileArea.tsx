import CloudUploadTwoTone from '@material-ui/icons/CloudUploadTwoTone'
import React, { useEffect, useState } from 'react'
import { useRepository } from '../hooks/use-repository'

export const DropFileArea: React.FunctionComponent<{
  uploadPath: string
  style?: React.CSSProperties
  uploadsetdata: () => void
  notificationControll: (onoff: boolean) => void
  setDragOver: (onoff: boolean) => void
  isDragOver: boolean
}> = props => {
  const repo = useRepository()
  const [scrollPosition, setScroll] = useState(0)

  useEffect(() => {
    window.onscroll = function() {
      setScroll(window.pageYOffset)
    }
  }, [])

  return (
    <div
      style={{
        display: 'flex',
        flexWrap: 'wrap',
        filter: props.isDragOver ? 'blur(1px)' : undefined,
        opacity: props.isDragOver ? 0.8 : 1,
        transition: 'opacity 300ms',
        ...props.style,
      }}
      onDragEnter={ev => {
        ev.stopPropagation()
        ev.preventDefault()
        props.setDragOver(true)
      }}
      onDragLeave={ev => {
        ev.stopPropagation()
        ev.preventDefault()
        props.setDragOver(false)
      }}
      onDragOver={ev => {
        ev.stopPropagation()
        ev.preventDefault()
        props.setDragOver(true)
      }}
      onDrop={async ev => {
        ev.stopPropagation()
        ev.preventDefault()
        props.setDragOver(false)
        await repo.upload.fromDropEvent({
          event: new DragEvent('drop', { dataTransfer: ev.dataTransfer }),
          binaryPropertyName: 'Binary',
          overwrite: true,
          createFolders: true,
          parentPath: props.uploadPath ? props.uploadPath : '',
          contentTypeName: 'Image',
        })
        props.uploadsetdata()
        props.notificationControll(true)
      }}>
      <div
        style={{
          width: 'auto',
          height: 'auto',
          backgroundColor: 'transparent',
          opacity: props.isDragOver ? 0.5 : 0,
          position: 'relative',
          transition: 'opacity 300ms',
        }}>
        <CloudUploadTwoTone
          style={{ width: '100%', position: 'absolute', marginTop: `${scrollPosition}px`, height: '250px' }}
        />
      </div>
      {props.children}
    </div>
  )
}
