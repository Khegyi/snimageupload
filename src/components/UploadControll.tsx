import React from 'react'
import CloudUpload from '@material-ui/icons/CloudUpload'
import { ConstantContent } from '@sensenet/client-core'
import { Fab, IconButton } from '@material-ui/core'
import CloseIcon from '@material-ui/icons/Close'
import Snackbar from '@material-ui/core/Snackbar'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import { useRepository } from '../hooks/use-repository'

interface UploadControllProps {
  uploadsetdata: () => void
  notificationControll: (onoff: boolean) => void
}
const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexGrow: 1,
      width: '100%',
      position: 'fixed',
      zIndex: 1,
    },
    input: {
      display: 'none',
    },
    button: {
      margin: theme.spacing(1),
    },
    title: {
      marginLeft: theme.spacing(2),
      flex: 1,
    },
    appBar: {
      width: '100%',
      marginRight: '0',
    },
  }),
)

export const UploadControll: React.FunctionComponent<UploadControllProps> = props => {
  const [open, setOpen] = React.useState(false)
  const repo = useRepository()
  const classes = useStyles()
  /**
   * Opens the Snackbar for succesfull upload
   */
  function handleClick() {
    setOpen(true)
  }
  /**
   * Closes the Snackbar
   * @param e any
   */
  function handleClose(_event: React.SyntheticEvent | React.MouseEvent, reason?: string) {
    if (reason === 'clickaway') {
      return
    }
    setOpen(false)
  }
  /**
   * Handle Uploaded File
   * @param e any
   */
  async function pickFile(e: any) {
    await repo.upload.fromFileList({
      binaryPropertyName: 'Binary',
      overwrite: true,
      parentPath: `${ConstantContent.PORTAL_ROOT.Path}/Content/IT/ImageLibrary`,
      file: e.target.files[0],
      contentTypeName: 'Image',
    })
    handleClick()
    props.uploadsetdata()
    props.notificationControll(true)
  }

  return (
    <div>
      <input
        accept="image/*"
        onChange={e => pickFile(e)}
        className={classes.input}
        id="contained-button-file"
        multiple
        type="file"
      />
      <label htmlFor="contained-button-file">
        <Fab variant="extended" title="Upload" component="span" className={classes.button}>
          <CloudUpload />
        </Fab>
      </label>
    </div>
  )
}
