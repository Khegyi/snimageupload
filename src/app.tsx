import React, { useEffect, useState } from 'react'
import { ConstantContent } from '@sensenet/client-core'
import { Image, User } from '@sensenet/default-content-types'
import { CssBaseline } from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import moment from 'moment'
import snLogo from './assets/sensenet_logo_transparent.png'
import { useRepository } from './hooks/use-repository'
import { AdvancedGridList } from './components/AdvancedGridList'
import { SimpleAppBar } from './components/SimpleAppBar'

interface SelectedImage {
  imgIndex: number
  imgPath: string
  imgTitle: string
  imgDescription: string
  imgAuthor: string
  imgAuthorAvatar: string
  imgCreationDate: string
  imgSize: string
  imgDownloadUrl: string
}
export const useStyles = makeStyles(theme => ({
  root: {
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'unset',
    overflow: 'visible',
    backgroundColor: theme.palette.background.paper,
  },
  content: {
    flexGrow: 1,
    padding: theme.spacing(3),
  },
}))
/**
 * The main entry point of your app. You can start h@cking from here ;)
 */
export const App: React.FunctionComponent = () => {
  const repo = useRepository()
  const [data, setData] = useState<Image[]>([])
  const [uploaddata, setUploaddata] = useState<boolean>(false)
  const [selectedimage, setSelectedimage] = React.useState<SelectedImage>({
    imgIndex: 0,
    imgPath: '',
    imgTitle: '',
    imgDescription: '',
    imgAuthor: '',
    imgAuthorAvatar: '',
    imgCreationDate: '',
    imgSize: '',
    imgDownloadUrl: '',
  })
  /**
   * Sets the UploadData
   */
  function setUploaddataFunction() {
    setUploaddata(true)
  }
  /**
   * Sets the Selected Image
   * @param {number} imageIndex Seletected number's index.
   */
  function getSelectedImage(imageIndex: number) {
    const selectedImage = data[imageIndex]
    const avatarUser = selectedImage.CreatedBy as User
    const avatarUserAvatarUrl = avatarUser.Avatar ? avatarUser.Avatar.Url : ''

    setSelectedimage({
      imgIndex: imageIndex,
      imgPath: repo.configuration.repositoryUrl + selectedImage.Path,
      imgTitle: selectedImage.DisplayName ? selectedImage.DisplayName : '',
      imgDescription: selectedImage.Description ? selectedImage.Description : '',
      imgAuthor: selectedImage.CreatedBy ? ((selectedImage.CreatedBy as User).FullName as string) : '',
      imgAuthorAvatar: avatarUserAvatarUrl as string,
      imgCreationDate: moment(new Date(selectedImage.CreationDate ? selectedImage.CreationDate : '')).format(
        'YYYY-MM-DD HH:mm:ss',
      ),
      imgSize: `${(selectedImage.Size ? selectedImage.Size / 1024 / 1024 : 0).toFixed(2)} MB`,
      imgDownloadUrl: selectedImage.Binary
        ? repo.configuration.repositoryUrl + selectedImage.Binary.__mediaresource.media_src
        : '',
    })
  }
  /**
   *  Close the Details View.
   */
  async function loadImages(): Promise<void> {
    const result = await repo.loadCollection<Image>({
      path: `${ConstantContent.PORTAL_ROOT.Path}/Content/IT/ImageLibrary`,
      oDataOptions: {
        select: [
          'Binary',
          'DisplayName',
          'Description',
          'CreationDate',
          'CreatedBy',
          'Height',
          'ModificationDate',
          'Size',
          'Width',
        ],
        expand: ['CreatedBy'],
      },
    })
    setData(result.d.results)
  }
  /**
   *  Close the Details View.
   */
  useEffect(() => {
    /**
     * Fetches the images from the repository.
     */
    console.log(uploaddata)
    if (uploaddata) {
      loadImages()
      setUploaddata(false)
    }
  }, [uploaddata])

  /**
   *  Close the Details View.
   */
  useEffect(() => {
    /**
     * Fetches the images from the repository.
     */
    loadImages()
  }, [repo])
  return (
    <div
      style={{
        display: 'flex',
        alignItems: 'center',
        width: '100%',
        height: '100%',
        flexDirection: 'column',
        backgroundPosition: 'center',
        backgroundRepeat: 'no-repeat',
        backgroundImage: `url(${snLogo})`,
        backgroundSize: 'auto',
      }}>
      <CssBaseline />
      <SimpleAppBar uploadsetdata={setUploaddataFunction} />
      <AdvancedGridList openFunction={getSelectedImage} imgList={data} />
    </div>
  )
}
