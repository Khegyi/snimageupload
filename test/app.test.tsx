import { mount, shallow } from 'enzyme'
import React from 'react'
import { act } from 'react-dom/test-utils'
import { Snackbar } from '@material-ui/core'
import { App } from '../src/app'
import { RepositoryContext } from '../src/context/repository-provider'
import { SimpleAppBar } from '../src/components/SimpleAppBar'
import { AdvancedGridList } from '../src/components/AdvancedGridList'
import { images } from './mocks/images'

describe('App Layout', () => {
  it('Matches snapshot', () => {
    const l = shallow(<App />)
    expect(l).toMatchSnapshot()
  })
  it('getselectedImage with result', async () => {
    const repository = {
      loadCollection: () => {
        return { d: { results: images } }
      },
      configuration: {
        repositoryUrl: 'url',
      },
    }
    let wrapper: any
    await act(async () => {
      wrapper = mount(
        <RepositoryContext.Provider value={repository as any}>
          <App />
        </RepositoryContext.Provider>,
      )
    })
    expect(
      wrapper
        .update()
        .find(AdvancedGridList)
        .prop('imgList'),
    ).not.toBe([])
  })
  it('getselectedImage without result', async () => {
    const repository = {
      loadCollection: () => {
        return { d: { results: [] } }
      },
      configuration: {
        repositoryUrl: 'url',
      },
    }
    let wrapper: any
    await act(async () => {
      wrapper = mount(
        <RepositoryContext.Provider value={repository as any}>
          <App />
        </RepositoryContext.Provider>,
      )
      expect(
        wrapper
          .update()
          .find(AdvancedGridList)
          .prop('imgList'),
      ).toStrictEqual([])
    })
  })
  it('Reload Images', async () => {
    const repository = {
      loadCollection: jest.fn(() => {
        return { d: { results: images } }
      }),
      configuration: {
        repositoryUrl: 'url',
      },
    }
    let wrapper: any
    await act(async () => {
      wrapper = mount(
        <RepositoryContext.Provider value={repository as any}>
          <App />
        </RepositoryContext.Provider>,
      )
    })
    expect(repository.loadCollection).toBeCalledTimes(1)
    await act(async () => {
      wrapper.find(SimpleAppBar).prop('uploadsetdata')()
    })
    expect(repository.loadCollection).toBeCalledTimes(2)
  })
  it('Show Notification', async () => {
    const repository = {
      loadCollection: () => {
        return { d: { results: [] } }
      },
      configuration: {
        repositoryUrl: 'url',
      },
    }
    let wrapper: any
    await act(async () => {
      wrapper = mount(
        <RepositoryContext.Provider value={repository as any}>
          <App />
        </RepositoryContext.Provider>,
      )
      await act(async () => {
        wrapper
          .update()
          .find(AdvancedGridList)
          .prop('notificationControll')(true)
      })
      expect(
        wrapper
          .update()
          .find(Snackbar)
          .prop('open'),
      ).toBe(true)
    })
  })
  it('Close Notification', async () => {
    const repository = {
      loadCollection: () => {
        return { d: { results: [] } }
      },
      configuration: {
        repositoryUrl: 'url',
      },
    }
    let wrapper: any
    await act(async () => {
      wrapper = mount(
        <RepositoryContext.Provider value={repository as any}>
          <App />
        </RepositoryContext.Provider>,
      )
      await act(async () => {
        wrapper
          .update()
          .find(Snackbar)
          .prop('onClose')()
      })
      expect(
        wrapper
          .update()
          .find(Snackbar)
          .prop('open'),
      ).toBe(false)
    })
  })
})
