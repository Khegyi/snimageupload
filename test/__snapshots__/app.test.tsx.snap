// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Layout Matches snapshot 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "backgroundImage": "url(sensenet_logo_transparent.png)",
      "backgroundPosition": "center",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "auto",
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
      "width": "100%",
    }
  }
>
  <WithStyles(ForwardRef(Snackbar))
    ContentProps={
      Object {
        "aria-describedby": "message-id",
      }
    }
    action={
      Array [
        <ForwardRef(WithStyles)
          aria-label="close"
          className="makeStyles-close-3"
          color="inherit"
          onClick={[Function]}
        >
          <Memo />
        </ForwardRef(WithStyles)>,
      ]
    }
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "bottom",
      }
    }
    autoHideDuration={6000}
    message={
      <span
        id="message-id"
      >
        Successful Upload
      </span>
    }
    onClose={[Function]}
    open={false}
  />
  <CssBaseline />
  <Component
    notificationControll={[Function]}
    uploadsetdata={[Function]}
  />
  <Component
    imgList={Array []}
    notificationControll={[Function]}
    uploadsetdata={[Function]}
  />
</div>
`;
